src/optimiser/taso/log.rs

```diff
@@ -47,36 +49,44 @@ impl<'w> TasoLogger<'w> {
     #[inline]
     pub fn log_processing_end<C: Debug>(
         &self,
-        circuit_count: usize,
+        circuits_processed: usize,
+        circuits_seen: Option<usize>,
         best_cost: C,
         needs_joining: bool,
         timeout: bool,
     ) {
         match timeout {
-            true => self.log("Optimisation finished (timeout)"),
-            false => self.log("Optimisation finished"),
+            true => self.log("Optimisation finished (timeout)."),
+            false => self.log("Optimisation finished."),
         };
-        self.log(format!("Tried {circuit_count} circuits"));
+        match circuits_seen {
+            Some(circuits_seen) => self.log(format!(
+                "Processed {circuits_processed} circuits (out of {circuits_seen} seen)."
+            )),
+            None => self.log(format!("Processed {circuits_processed} circuits.")),
+        }
         self.log(format!("---- END RESULT: {:?} ----", best_cost));
         if needs_joining {
-            self.log("Joining worker threads");
+            self.log("Joining worker threads.");
         }
     }
 
     /// Log the progress of the optimisation.
     #[inline(always)]
     pub fn log_progress(
         &mut self,
-        circ_cnt: usize,
+        circuits_processed: usize,
         workqueue_len: Option<usize>,
         seen_hashes: usize,
     ) {
-        if circ_cnt % 1000 == 0 {
-            self.progress(format!("{circ_cnt} circuits..."));
+        if circuits_processed > self.last_circ_processed && circuits_processed % 1000 == 0 {
```

Comment:
Can't you simplify this to `if circuits_processed > self.last_circ_processed + 1000`? Then we might also get away with less `log_counts` calls.